src/account/operations/output_finder.rs

```diff
@@ -8,37 +8,55 @@ use iota_client::secret::GenerateAddressMetadata;
 use crate::account::{
     handle::AccountHandle,
     operations::{address_generation::AddressGenerationOptions, syncing::SyncOptions},
+    types::AddressWithUnspentOutputs,
 };
 
 impl AccountHandle {
     /// Search addresses with unspent outputs
-    /// `address_gap_limit` defines how many addresses without outputs will be checked in each account, if an address
-    /// has outputs, the counter is reset
+    /// `address_gap_limit`: The number of addresses to search for, after the last address with unspent outputs
     /// Addresses that got crated during this operation and have a higher key_index than the latest one with outputs,
     /// will be removed again, to keep the account size smaller
     pub(crate) async fn search_addresses_with_outputs(
         self: &AccountHandle,
-        address_gap_limit: u32,
+        mut address_gap_limit: u32,
         mut sync_options: Option<SyncOptions>,
     ) -> crate::Result<usize> {
         log::debug!("[search_addresses_with_outputs]");
 
-        // store the length so we can remove addresses with higher indexes later if they don't have outputs
+        let mut address_gap_limit_internal = address_gap_limit;
+
+        // store the current index, so we can remove new addresses with higher indexes later again, if they don't have
+        // outputs
         let (highest_public_address_index, highest_internal_address_index) = {
             let account = self.read().await;
-            let highest_public_address_index = match account.public_addresses.last() {
-                Some(a) => a.key_index,
-                None => 0,
-            };
-            let highest_internal_address_index = match account.internal_addresses.last() {
-                Some(a) => a.key_index,
-                None => 0,
-            };
-            (highest_public_address_index, highest_internal_address_index)
+            (
+                account
+                    .public_addresses
+                    .last()
+                    .map(|a| a.key_index)
+                    .expect("account needs to have a public address"),
+                account.internal_addresses.last().map(|a| a.key_index),
+            )
         };
 
         let mut latest_outputs_count = 0;
         loop {
+            // Also needs to be in the loop so it gets updated every round for internal use without modifying the values
+            // outside
+            let (highest_public_address_index, highest_internal_address_index) = {
+                let account = self.read().await;
+                (
+                    account
+                        .public_addresses
+                        .last()
+                        .map(|a| a.key_index)
+                        .expect("account needs to have a public address"),
+                    account.internal_addresses.last().map(|a| a.key_index),
```

Comment:
Is there no other way at all to do that? Because to full iterations inside an iteration is not idea

Reply:
What do you mean with full iteration?

Reply:
Oh this is `Vec::last` and not `Iterator::last`, ignore me 🙈